File: src/internal/server/garapons.go
package server

import (
	"database/sql"
	"errors"
	"net/http"
	"strconv"
	"strings"

	"app-suite/internal/model"
	"app-suite/internal/store"
)

// ── Garapon admin (list + detail + CRUD + drawing links) ────────────────────
//
// A garapon is a festival lottery drum (see model.Garapon). Admins manage it like
// a raffle — create/edit/close — but instead of public sign-up, each player gets a
// private tokenized link (a GaraponPlayer) with a draw allowance. The public draw
// endpoints below need no auth; the token is the capability.

// handleGaraponsList returns every garapon (admin only).
//
//	Endpoint:  GET /api/garapons
//	Auth:      admin, or a user granted festival-garapon
//	Response:  {"garapons": [...]}
func (s *Server) handleGaraponsList(w http.ResponseWriter, r *http.Request) {
	if !s.requirePermission(w, r, permFestivalGarapon) {
		return
	}
	garapons, err := s.store.ListGarapons()
	if err != nil {
		writeInternalError(w, "list garapons", err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"garapons": garapons})
}

// handleGaraponDetail returns a single garapon with its prizes, drawing links,
// and full draw log (admin only).
//
//	Endpoint:  GET /api/garapons/{id}
//	Auth:      admin, or a user granted festival-garapon
//	Response:  {"garapon": Garapon, "players": [...], "draws": [...]}
func (s *Server) handleGaraponDetail(w http.ResponseWriter, r *http.Request) {
	if !s.requirePermission(w, r, permFestivalGarapon) {
		return
	}
	id, err := strconv.ParseInt(r.PathValue("id"), 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "Invalid garapon ID")
		return
	}
	garapon, err := s.store.GetGarapon(id)
	if err != nil {
		writeInternalError(w, "get garapon", err)
		return
	}
	if garapon == nil {
		writeError(w, http.StatusNotFound, "Garapon not found")
		return
	}
	players, err := s.store.ListGaraponPlayers(id)
	if err != nil {
		writeInternalError(w, "list garapon players", err)
		return
	}
	draws, err := s.store.ListGaraponDraws(id)
	if err != nil {
		writeInternalError(w, "list garapon draws", err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"garapon": garapon, "players": players, "draws": draws})
}

// garaponRequest is the JSON body for POST /api/garapons.
// Action: "create", "update", "delete", or "set_status".
type garaponRequest struct {
	Action          string               `json:"action"`
	ID              int64                `json:"id"`
	Title           string               `json:"title"`
	Details         string               `json:"details"`
	GrandPrizeImage string               `json:"grand_prize_image"`
	Status          string               `json:"status"`
	Prizes          []model.GaraponPrize `json:"prizes"`
}

// sanitizeGaraponPrizes trims/normalizes incoming prize rows: it drops blank rows,
// defaults an empty ball color, floors negative rates at 0, and enforces exactly
// one grand prize (defaulting the first when none is flagged). Returns a non-empty
// user-facing message when no usable prizes remain or more than one grand prize is
// flagged (so the handler can pass it straight to writeError).
func sanitizeGaraponPrizes(in []model.GaraponPrize) (prizes []model.GaraponPrize, errMsg string) {
	out := make([]model.GaraponPrize, 0, len(in))
	grandCount := 0
	for _, p := range in {
		name := strings.TrimSpace(p.Name)
		if name == "" {
			continue
		}
		color := strings.TrimSpace(p.BallColor)
		if color == "" {
			color = "#cccccc"
		}
		rate := p.Rate
		if rate < 0 {
			rate = 0
		}
		if p.IsGrand {
			grandCount++
		}
		out = append(out, model.GaraponPrize{Name: name, BallColor: color, Rate: rate, IsGrand: p.IsGrand})
	}
	if len(out) == 0 {
		return nil, "At least one prize is required"
	}
	if grandCount == 0 {
		out[0].IsGrand = true
	} else if grandCount > 1 {
		return nil, "Only one prize can be the grand prize"
	}
	return out, ""
}

// handleGaraponsAction processes garapon CRUD + status changes.
//
//	Endpoint:  POST /api/garapons
//	Auth:      admin, or a user granted festival-garapon
//	Request:   {"action": "create"|"update"|"delete"|"set_status", ...}
func (s *Server) handleGaraponsAction(w http.ResponseWriter, r *http.Request) {
	if !s.requirePermission(w, r, permFestivalGarapon) {
		return
	}
	req, err := readJSON[garaponRequest](w, r)
	if err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}

	switch req.Action {
	case "create":
		title := strings.TrimSpace(req.Title)
		if title == "" {
			writeError(w, http.StatusBadRequest, "Title is required")
			return
		}
		prizes, msg := sanitizeGaraponPrizes(req.Prizes)
		if msg != "" {
			writeError(w, http.StatusBadRequest, msg)
			return
		}
		garapon := &model.Garapon{
			Title:           title,
			Details:         req.Details,
			GrandPrizeImage: req.GrandPrizeImage,
			Prizes:          prizes,
		}
		id, err := s.store.CreateGarapon(garapon)
		if err != nil {
			writeInternalError(w, "create garapon", err)
			return
		}
		garapon.ID = id
		garapon.Status = "open"
		writeJSON(w, http.StatusCreated, map[string]any{"garapon": garapon})

	case "update":
		if req.ID <= 0 {
			writeError(w, http.StatusBadRequest, "Garapon id is required")
			return
		}
		title := strings.TrimSpace(req.Title)
		if title == "" {
			writeError(w, http.StatusBadRequest, "Title is required")
			return
		}
		prizes, msg := sanitizeGaraponPrizes(req.Prizes)
		if msg != "" {
			writeError(w, http.StatusBadRequest, msg)
			return
		}
		garapon := &model.Garapon{
			ID:              req.ID,
			Title:           title,
			Details:         req.Details,
			GrandPrizeImage: req.GrandPrizeImage,
			Prizes:          prizes,
		}
		if err := s.store.UpdateGarapon(garapon); err != nil {
			writeInternalError(w, "update garapon", err)
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"ok": true})

	case "delete":
		if req.ID <= 0 {
			writeError(w, http.StatusBadRequest, "Garapon id is required")
			return
		}
		deleted, err := s.store.DeleteGarapon(req.ID)
		if err != nil {
			writeInternalError(w, "delete garapon", err)
			return
		}
		// The grand-prize image is managed centrally on System → Images (the
		// "Garapon" category), so the file is left intact — it may be reused.
		writeJSON(w, http.StatusOK, map[string]any{"deleted": deleted})

	case "set_status":
		if req.ID <= 0 {
			writeError(w, http.StatusBadRequest, "Garapon id is required")
			return
		}
		if req.Status != "open" && req.Status != "closed" {
			writeError(w, http.StatusBadRequest, "Status must be \"open\" or \"closed\"")
			return
		}
		if err := s.store.SetGaraponStatus(req.ID, req.Status); err != nil {
			writeInternalError(w, "set garapon status", err)
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"ok": true, "status": req.Status})

	default:
		writeError(w, http.StatusBadRequest, "Invalid action. Use: create, update, delete, set_status")
	}
}

// garaponPlayersRequest is the JSON body for POST /api/garapons/{id}/players.
// Action: "create_player" or "delete_player".
type garaponPlayersRequest struct {
	Action     string `json:"action"`
	PlayerID   int64  `json:"player_id"`
	PlayerName string `json:"player_name"`
	MaxDraws   int    `json:"max_draws"`
}

// handleGaraponPlayers creates or deletes a garapon's drawing links.
//
//	Endpoint:  POST /api/garapons/{id}/players
//	Auth:      admin, or a user granted festival-garapon
//	Request:   {"action":"create_player","player_name":"...","max_draws":5}
//	           {"action":"delete_player","player_id":1}
func (s *Server) handleGaraponPlayers(w http.ResponseWriter, r *http.Request) {
	if !s.requirePermission(w, r, permFestivalGarapon) {
		return
	}
	garaponID, err := strconv.ParseInt(r.PathValue("id"), 10, 64)
	if err != nil {
		writeError(w, http.StatusBadRequest, "Invalid garapon ID")
		return
	}
	req, err := readJSON[garaponPlayersRequest](w, r)
	if err != nil {
		writeError(w, http.StatusBadRequest, "Invalid JSON")
		return
	}

	switch req.Action {
	case "create_player":
		name := strings.TrimSpace(req.PlayerName)
		if name == "" {
			writeError(w, http.StatusBadRequest, "Player name is required")
			return
		}
		maxDraws := req.MaxDraws
		if maxDraws < 1 {
			maxDraws = 1
		}
		garapon, err := s.store.GetGarapon(garaponID)
		if err != nil {
			writeInternalError(w, "get garapon for player", err)
			return
		}
		if garapon == nil {
			writeError(w, http.StatusNotFound, "Garapon not found")
			return
		}
		player, err := s.store.CreateGaraponPlayer(garaponID, name, maxDraws)
		if err != nil {
			writeInternalError(w, "create garapon player", err)
			return
		}
		writeJSON(w, http.StatusCreated, map[string]any{"player": player})

	case "delete_player":
		if req.PlayerID <= 0 {
			writeError(w, http.StatusBadRequest, "Player id is required")
			return
		}
		existing, err := s.store.GetGaraponPlayerByID(req.PlayerID)
		if err != nil {
			writeInternalError(w, "get garapon player", err)
			return
		}
		// A closed garapon can be cleaned up: any link may be deleted, and its
		// draws stay in the log (garapon_draws.player_id is ON DELETE SET NULL).
		// While the garapon is open, a link that has already drawn can't be deleted.
		garapon, err := s.store.GetGarapon(garaponID)
		if err != nil {
			writeInternalError(w, "get garapon for delete player", err)
			return
		}
		closed := garapon != nil && garapon.Status == "closed"
		if existing != nil && existing.DrawsUsed > 0 && !closed {
			writeError(w, http.StatusConflict,
				"This player has already drawn and can't be deleted while the garapon is open")
			return
		}
		deleted, err := s.store.DeleteGaraponPlayer(req.PlayerID, closed)
		if err != nil {
			writeInternalError(w, "delete garapon player", err)
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"deleted": deleted})

	default:
		writeError(w, http.StatusBadRequest, "Invalid action. Use: create_player, delete_player")
	}
}

// ── Garapon public (tokenized player view + draw) ───────────────────────────

// publicGarapon is the trimmed garapon shape sent to players: it carries the
// prizes (names + ball colors + which is grand) but never the appearance rates —
// the odds stay admin-only.
type publicGarapon struct {
	ID              int64                `json:"id"`
	Title           string               `json:"title"`
	Details         string               `json:"details"`
	GrandPrizeImage string               `json:"grand_prize_image"`
	Status          string               `json:"status"`
	Prizes          []model.GaraponPrize `json:"prizes"`
}

// toPublicGarapon copies a garapon for the player view, zeroing each prize's Rate
// so the configured odds aren't exposed.
func toPublicGarapon(g *model.Garapon) publicGarapon {
	prizes := make([]model.GaraponPrize, len(g.Prizes))
	for i, p := range g.Prizes {
		p.Rate = 0
		prizes[i] = p
	}
	return publicGarapon{
		ID:              g.ID,
		Title:           g.Title,
		Details:         g.Details,
		GrandPrizeImage: g.GrandPrizeImage,
		Status:          g.Status,
		Prizes:          prizes,
	}
}

// publicPlayer is the trimmed player shape sent to the public view (no token —
// the caller already holds it in the URL).
type publicPlayer struct {
	PlayerName string `json:"player_name"`
	MaxDraws   int    `json:"max_draws"`
	DrawsUsed  int    `json:"draws_used"`
}

// loadGaraponByToken resolves a player token to its player + garapon, writing the
// 404 itself when the token is unknown. Returns (player, garapon, ok).
func (s *Server) loadGaraponByToken(w http.ResponseWriter, token string) (*model.GaraponPlayer, *model.Garapon, bool) {
	player, err := s.store.GetGaraponPlayerByToken(token)
	if err != nil {
		writeInternalError(w, "get garapon player by token", err)
		return nil, nil, false
	}
	if player == nil {
		writeError(w, http.StatusNotFound, "Drawing link not found")
		return nil, nil, false
	}
	garapon, err := s.store.GetGarapon(player.GaraponID)
	if err != nil {
		writeInternalError(w, "get garapon for token", err)
		return nil, nil, false
	}
	if garapon == nil {
		writeError(w, http.StatusNotFound, "Garapon not found")
		return nil, nil, false
	}
	return player, garapon, true
}

// handleGaraponPublic returns the player-facing garapon view for a drawing token:
// the garapon (no odds), the player's name + draw allowance/usage, and their own
// draw history.
//
//	Endpoint:  GET /api/garapon/{token}
//	Auth:      public (the token is the capability)
//	Response:  {"garapon": {...}, "player": {...}, "draws": [...]}
func (s *Server) handleGaraponPublic(w http.ResponseWriter, r *http.Request) {
	token := strings.TrimSpace(r.PathValue("token"))
	player, garapon, ok := s.loadGaraponByToken(w, token)
	if !ok {
		return
	}
	draws, err := s.store.ListPlayerDraws(player.ID)
	if err != nil {
		writeInternalError(w, "list player draws", err)
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"garapon": toPublicGarapon(garapon),
		"player":  publicPlayer{PlayerName: player.PlayerName, MaxDraws: player.MaxDraws, DrawsUsed: player.DrawsUsed},
		"draws":   draws,
	})
}

// handleGaraponDraw performs one authoritative draw for a token. The store
// re-checks the open status + remaining-draw cap atomically and weighted-picks the
// prize, so the client can't bias the odds or exceed the allowance.
//
//	Endpoint:  POST /api/garapon/{token}/draw
//	Auth:      public (the token is the capability)
//	Response:  {"draw": GaraponDraw, "draws_used": int, "max_draws": int}
func (s *Server) handleGaraponDraw(w http.ResponseWriter, r *http.Request) {
	token := strings.TrimSpace(r.PathValue("token"))
	player, _, ok := s.loadGaraponByToken(w, token)
	if !ok {
		return
	}

	draw, err := s.store.RecordGaraponDraw(player.ID)
	switch {
	case errors.Is(err, store.ErrGaraponClosed):
		writeError(w, http.StatusBadRequest, "This garapon is closed")
		return
	case errors.Is(err, store.ErrGaraponNoDraws):
		writeError(w, http.StatusConflict, "No draws remaining")
		return
	case errors.Is(err, store.ErrGaraponNoPrizes):
		writeError(w, http.StatusBadRequest, "This garapon has no prizes configured")
		return
	case errors.Is(err, sql.ErrNoRows):
		writeError(w, http.StatusNotFound, "Drawing link not found")
		return
	case err != nil:
		writeInternalError(w, "record garapon draw", err)
		return
	}

	// Reload the player for the fresh used-count so the client can update the
	// remaining-draws display without a second request.
	updated, err := s.store.GetGaraponPlayerByID(player.ID)
	usedCount := player.DrawsUsed + 1
	maxDraws := player.MaxDraws
	if err == nil && updated != nil {
		usedCount = updated.DrawsUsed
		maxDraws = updated.MaxDraws
	}
	writeJSON(w, http.StatusOK, map[string]any{"draw": draw, "draws_used": usedCount, "max_draws": maxDraws})
}

File: frontend/src/stores/garapons.ts
/**
 * Garapons store: admin management (CRUD, prizes, per-player drawing links, draw
 * log) plus the public token-based player view + authoritative draw.
 *
 * Structurally a leaner cousin of the raffles store — a garapon has no ticket
 * sign-up or cost; instead an admin issues each player a tokenized link with a
 * draw allowance, and the server picks each prize. The grand-prize image is
 * picked from the "Garapon" image category, exactly like raffle prize images.
 */
import { defineStore } from 'pinia'
import { computed, ref } from 'vue'
import { endpoints } from '@/lib/endpoints'
import type {
  Garapon,
  GaraponDraw,
  GaraponDrawResponse,
  GaraponForm,
  GaraponPlayer,
  GaraponPrize,
  GaraponPrizeForm,
} from '@/types/api'
import { useUiStore } from './ui'
import { useImagesStore, IMAGE_DIR_GARAPONS } from './images'

/** A sensible default ball color for a fresh prize row (a festival gold). */
const DEFAULT_BALL_COLOR = '#e5b53f'

/** A fresh prize row for the editor (the first/grand row defaults to a higher weight). */
function blankPrize(rate: number, isGrand = false): GaraponPrizeForm {
  return { name: '', ball_color: DEFAULT_BALL_COLOR, rate, is_grand: isGrand }
}

export const useGaraponsStore = defineStore('garapons', () => {
  const ui = useUiStore()

  // ── Admin state ──────────────────────────────────────────────────────────
  const garapons = ref<Garapon[]>([])
  const selectedGarapon = ref<Garapon | null>(null)
  const garaponPlayers = ref<GaraponPlayer[]>([])
  const garaponDraws = ref<GaraponDraw[]>([])
  const garaponForm = ref<GaraponForm | null>(null)
  /** Reusable grand-prize images (the "Garapon" category on System → Images). */
  const grandPrizeImages = ref<string[]>([])
  /** Admin "generate drawing" form (issue a new per-player link). */
  const playerAdd = ref<{ playerName: string; maxDraws: number }>({ playerName: '', maxDraws: 1 })

  const garaponsLoading = ref(false)
  const detailLoading = ref(false)
  const savingGarapon = ref(false)
  const creatingPlayer = ref(false)

  // ── Public (player view) state ───────────────────────────────────────────
  const publicGarapon = ref<Garapon | null>(null)
  const publicPlayer = ref<{ player_name: string; max_draws: number; draws_used: number } | null>(
    null,
  )
  const publicDraws = ref<GaraponDraw[]>([])
  /** The most recent win (drives the "congratulations" banner). */
  const lastWin = ref<GaraponDraw | null>(null)
  const publicLoading = ref(false)
  const drawing = ref(false)

  // ── Computed ─────────────────────────────────────────────────────────────
  const openGarapons = computed(() => garapons.value.filter((g) => g.status === 'open'))
  const closedGarapons = computed(() => garapons.value.filter((g) => g.status === 'closed'))

  /** The flagged grand prize of the currently-viewed public garapon (if any). */
  const grandPrize = computed<GaraponPrize | null>(
    () => publicGarapon.value?.prizes?.find((p) => p.is_grand) ?? null,
  )
  /** The non-grand prizes (the "other possible prizes" table). */
  const otherPrizes = computed<GaraponPrize[]>(
    () => publicGarapon.value?.prizes?.filter((p) => !p.is_grand) ?? [],
  )
  /** Draws the public player has left. */
  const drawsRemaining = computed(() =>
    publicPlayer.value ? Math.max(0, publicPlayer.value.max_draws - publicPlayer.value.draws_used) : 0,
  )
  /** Whether the public player can draw right now. */
  const canDraw = computed(
    () => publicGarapon.value?.status === 'open' && drawsRemaining.value > 0 && !drawing.value,
  )

  // ── Admin: load ──────────────────────────────────────────────────────────
  async function loadGarapons(): Promise<void> {
    garaponsLoading.value = true
    try {
      const data = await endpoints.garapons.list()
      garapons.value = data.garapons || []
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    } finally {
      garaponsLoading.value = false
    }
  }

  async function loadGaraponDetail(id: number): Promise<void> {
    detailLoading.value = true
    try {
      const data = await endpoints.garapons.detail(id)
      selectedGarapon.value = data.garapon
      garaponPlayers.value = data.players || []
      garaponDraws.value = data.draws || []
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    } finally {
      detailLoading.value = false
    }
  }

  /** Admin: open a garapon's detail view. */
  function viewGarapon(g: Garapon): void {
    selectedGarapon.value = g
    garaponPlayers.value = []
    garaponDraws.value = []
    resetPlayerAdd()
    loadGaraponDetail(g.id)
  }

  function resetPlayerAdd(): void {
    playerAdd.value = { playerName: '', maxDraws: 1 }
  }

  /** Loads the reusable grand-prize images (the "Garapon" category) for the picker. */
  async function loadGrandPrizeImages(): Promise<void> {
    try {
      const images = useImagesStore()
      await images.loadImages(IMAGE_DIR_GARAPONS)
      grandPrizeImages.value = (images.imagesByDir[IMAGE_DIR_GARAPONS] || []).map((i) => i.path)
    } catch {
      /* non-fatal: the picker just shows nothing */
    }
  }

  // ── Admin: form ──────────────────────────────────────────────────────────
  function newGaraponForm(): void {
    garaponForm.value = {
      id: 0,
      title: '',
      details: '',
      grand_prize_image: '',
      prizes: [blankPrize(50, true)],
    }
  }

  function editGaraponForm(g: Garapon): void {
    const prizes = (g.prizes || []).map((p) => ({
      name: p.name,
      ball_color: p.ball_color || DEFAULT_BALL_COLOR,
      rate: p.rate,
      is_grand: p.is_grand,
    }))
    // Guarantee exactly one grand prize for the radio group.
    if (prizes.length && !prizes.some((p) => p.is_grand)) prizes[0].is_grand = true
    garaponForm.value = {
      id: g.id,
      title: g.title,
      details: g.details,
      grand_prize_image: g.grand_prize_image,
      prizes: prizes.length ? prizes : [blankPrize(50, true)],
    }
  }

  function cancelGaraponForm(): void {
    garaponForm.value = null
  }

  /** Adds a blank prize row to the form. */
  function addPrizeRow(): void {
    garaponForm.value?.prizes.push(blankPrize(10))
  }

  /** Removes a prize row; if it was the grand prize, promotes the first remaining row. */
  function removePrizeRow(index: number): void {
    const f = garaponForm.value
    if (!f || f.prizes.length <= 1) return
    const wasGrand = f.prizes[index].is_grand
    f.prizes.splice(index, 1)
    if (wasGrand && !f.prizes.some((p) => p.is_grand)) f.prizes[0].is_grand = true
  }

  /** Marks one prize row as the grand prize (radio-style, single selection). */
  function setGrandPrize(index: number): void {
    const f = garaponForm.value
    if (!f) return
    f.prizes.forEach((p, i) => (p.is_grand = i === index))
  }

  /** Saves the garapon form. Returns true on success (caller navigates). */
  async function saveGarapon(): Promise<boolean> {
    const f = garaponForm.value
    if (!f) return false
    if (!f.title.trim()) {
      ui.notify('Title is required', 'error')
      return false
    }
    const named = f.prizes.filter((p) => p.name.trim())
    if (!named.length) {
      ui.notify('Add at least one prize with a name', 'error')
      return false
    }
    savingGarapon.value = true
    try {
      const payload = { ...f, prizes: named }
      if (f.id) {
        await endpoints.garapons.update(payload)
        ui.notify('Garapon updated', 'success')
      } else {
        await endpoints.garapons.create(payload)
        ui.notify('Garapon created', 'success')
      }
      garaponForm.value = null
      await loadGarapons()
      return true
    } catch (e) {
      ui.notify((e as Error).message, 'error')
      return false
    } finally {
      savingGarapon.value = false
    }
  }

  async function deleteGarapon(id: number): Promise<void> {
    if (
      !(await ui.confirm('Delete this garapon and all its drawing links and results?', {
        title: 'Delete garapon',
        confirmText: 'Delete',
      }))
    )
      return
    try {
      await endpoints.garapons.delete(id)
      garapons.value = garapons.value.filter((g) => g.id !== id)
      if (selectedGarapon.value?.id === id) selectedGarapon.value = null
      ui.notify('Garapon deleted', 'info')
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    }
  }

  async function setGaraponStatus(id: number, status: 'open' | 'closed'): Promise<void> {
    try {
      await endpoints.garapons.setStatus(id, status)
      if (selectedGarapon.value?.id === id) selectedGarapon.value.status = status
      const inList = garapons.value.find((g) => g.id === id)
      if (inList) inList.status = status
      ui.notify(status === 'closed' ? 'Garapon closed' : 'Garapon reopened', 'success')
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    }
  }

  // ── Admin: drawing links ─────────────────────────────────────────────────
  async function createPlayer(): Promise<void> {
    if (!selectedGarapon.value) return
    const name = playerAdd.value.playerName.trim()
    if (!name) {
      ui.notify('Player name is required', 'error')
      return
    }
    const max = Math.max(1, Math.floor(Number(playerAdd.value.maxDraws) || 1))
    creatingPlayer.value = true
    try {
      const data = await endpoints.garapons.createPlayer(selectedGarapon.value.id, {
        player_name: name,
        max_draws: max,
      })
      // Copy the new link straight to the clipboard so the admin can paste it to
      // the player. The URL is too long for the toast, so we only confirm the
      // action; do the copy before the detail reload to stay within the click's
      // activation window, and fall back gracefully if the browser blocks it.
      let copied = false
      try {
        await navigator.clipboard.writeText(playerLinkUrl(data.player))
        copied = true
      } catch {
        /* clipboard blocked (insecure context / permissions) — the per-row Copy link button still works */
      }
      ui.notify(
        copied ? 'Drawing link created and copied to clipboard' : 'Drawing link created',
        'success',
      )
      resetPlayerAdd()
      await loadGaraponDetail(selectedGarapon.value.id)
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    } finally {
      creatingPlayer.value = false
    }
  }

  async function deletePlayer(player: GaraponPlayer): Promise<void> {
    if (!selectedGarapon.value) return
    // Deleting a link keeps its draws in the log (they detach, not delete), so
    // reassure the admin when the player has already drawn.
    const msg =
      player.draws_used > 0
        ? `Delete the drawing link for ${player.player_name}? Their ${player.draws_used} draw${
            player.draws_used === 1 ? '' : 's'
          } stay in the draw log.`
        : `Delete the drawing link for ${player.player_name}?`
    if (!(await ui.confirm(msg, { title: 'Delete drawing link', confirmText: 'Delete' }))) return
    try {
      await endpoints.garapons.deletePlayer(selectedGarapon.value.id, player.id)
      garaponPlayers.value = garaponPlayers.value.filter((p) => p.id !== player.id)
      ui.notify('Drawing link deleted', 'info')
    } catch (e) {
      ui.notify((e as Error).message, 'error')
    }
  }

  /** A player's full public drawing link (origin + tokenized path). */
  function playerLinkUrl(player: GaraponPlayer): string {
    return `${window.location.origin}/garapon/${player.token}`
  }

  /** Builds a player's full public link and copies it to the clipboard. */
  async function copyPlayerLink(player: GaraponPlayer): Promise<void> {
    const url = playerLinkUrl(player)
    try {
      await navigator.clipboard.writeText(url)
      ui.notify('Link copied to clipboard', 'success')
    } catch {
      // Clipboard blocked (insecure context / permissions) — surface the URL.
      ui.notify(url, 'info')
    }
  }

  // ── Public: player view + draw ───────────────────────────────────────────
  function resetPublic(): void {
    publicGarapon.value = null
    publicPlayer.value = null
    publicDraws.value = []
    lastWin.value = null
  }

  /** Loads a player's garapon by token. Returns true on success. */
  async function loadByToken(token: string): Promise<boolean> {
    resetPublic()
    publicLoading.value = true
    try {
      const data = await endpoints.garapon.get(token)
      publicGarapon.value = data.garapon
      publicPlayer.value = data.player
      publicDraws.value = data.draws || []
      return true
    } catch {
      return false
    } finally {
      publicLoading.value = false
    }
  }

  /**
   * Performs one authoritative draw via the token. Returns the server response
   * (the won prize + fresh counts) so the wheel can animate to it, or null on
   * error. State is committed separately via `commitDraw` once the ball lands, so
   * the record/counts reveal in sync with the animation.
   */
  async function draw(token: string): Promise<GaraponDrawResponse | null> {
    if (drawing.value) return null
    drawing.value = true
    try {
      return await endpoints.garapon.draw(token)
    } catch (e) {
      ui.notify((e as Error).message, 'error')
      return null
    } finally {
      drawing.value = false
    }
  }

  /** Commits a drawn result to the visible record + remaining count + banner. */
  function commitDraw(resp: GaraponDrawResponse): void {
    publicDraws.value = [...publicDraws.value, resp.draw]
    if (publicPlayer.value) {
      publicPlayer.value.draws_used = resp.draws_used
      publicPlayer.value.max_draws = resp.max_draws
    }
    lastWin.value = resp.draw
  }

  return {
    // admin state
    garapons,
    selectedGarapon,
    garaponPlayers,
    garaponDraws,
    garaponForm,
    grandPrizeImages,
    playerAdd,
    garaponsLoading,
    detailLoading,
    savingGarapon,
    creatingPlayer,
    // public state
    publicGarapon,
    publicPlayer,
    publicDraws,
    lastWin,
    publicLoading,
    drawing,
    // computed
    openGarapons,
    closedGarapons,
    grandPrize,
    otherPrizes,
    drawsRemaining,
    canDraw,
    // admin actions
    loadGarapons,
    loadGaraponDetail,
    viewGarapon,
    loadGrandPrizeImages,
    newGaraponForm,
    editGaraponForm,
    cancelGaraponForm,
    addPrizeRow,
    removePrizeRow,
    setGrandPrize,
    saveGarapon,
    deleteGarapon,
    setGaraponStatus,
    createPlayer,
    deletePlayer,
    copyPlayerLink,
    // public actions
    resetPublic,
    loadByToken,
    draw,
    commitDraw,
  }
})
